Add config to enable notify_push reverse proxy

Pilzinsel64 opened the pull request:
Fix #1655

I used a few minutes and tried to implement a simple reverse proxy for the notiy_push component as a first step (see linked issue). It's probably all we can do. But this will already help a lot people to use the notify_push component **with encrypted https**!

In a later step we can maybe try to install and setup notify_push as additional component (maybe enabled by default). But 1. we would need to know the domain for the setup and 2. I have not the knowledge about the most programing stuff here. xP

## notify_push configuration

By default, the snap does not enable notify_push reverse proxy. To enable it, run:

    sudo snap set nextcloud http.notify-push-reverse-proxy=true

To disable it, run:

    sudo snap set nextcloud http.notify-push-reverse-proxy=false

## Some questions

1. @kyrofa What should we do with test? It is not possible to test the reverse proxy afaik. If there were a fully configured and running notify_push component then an empty page should be returned. when accessing https://cloud.mydomain.com/push/

2. What about the apache configuration? Is that correct on that place here or should I move it to `ssl.conf` to the `VirtualHost` section for https connection?

3. If the tests don't fail, would you @kyrofa please kick-off a build to `latest/beta/pr-2490`, so we can test that a bit?

4. Should I add an entry to the `README.md` file or just adjust the [wiki FAQ entry](https://github.com/nextcloud-snap/nextcloud-snap/wiki/FAQ's#q-how-to-install-files-hpb-client-push) I made a few weeks ago? In my opinion adjusting the wiki entry is enough as the config value is very special compared to the other ones. On the other side ont he README there is every single config option listed at the moment.

## Todo

- [x] Make a new [wiki page](https://github.com/nextcloud-snap/nextcloud-snap/wiki/Configure-HPB-client-push-for-Nextcloud-snap)
- [x] Make a [PR to notify_push repo](https://github.com/nextcloud/notify_push/pull/345)
- [ ] Remove the old [wiki FAQ entry](https://github.com/nextcloud-snap/nextcloud-snap/wiki/FAQ's#q-how-to-install-files-hpb-client-push) _(after merge!)_

kyrofa commented:
Hmm, yeah I'm going to need to get a lot smarter on notify_push in order to properly answer your questions. Except for this one:

> If the tests don't fail, would you @kyrofa please kick-off a build to latest/beta/pr-2490, so we can test that a bit?

Yes of course.

pachulo commented:
Thanks for this @Pilzinsel64 ! Did you try it?

Pilzinsel64 commented:
Not Yet, I had not the time before this week to setup a VM environment and build it myself without SSL of course. My initial plan was to let it install and build during this day, hope I can get to it, still have vocation at home. :)
If I am not able to test in VM environment i would need to install dependencies and build on my prod Maschine but currently I try to avoid that as it is a mostly fresh installed system. I would test it on my prod as soon as a build is on snapcraft, as there and then I can check it very chickly if it works, within two mins.

Pilzinsel64 commented:
Alright, after a bit trouble at build again, I got the snap working!

The changes made in the PR seems to work, see the video. The response for `http://mydomain.com/push/` and `http://mydomain.com/push/ws/` when the reverse proxy is disabled is `404`. When the reverse proxy is enabled, the response is `508`. This looks good for me.

However, I'm not able to setup notify_push without a real domain and SSL yet.
As I have nofiy_push pre-configured on my prod server, I gently wanna ask again for builds at `latest/beta/pr-2490`. :)

https://github.com/nextcloud-snap/nextcloud-snap/assets/23138465/7a78c38b-cee2-4b37-9d81-f4be743647b6

Pilzinsel64 commented:
It failed. 😢
But the artifact has been attached sucessfully.
Misterious, it has four PR numbers listed at the two previous tasks and fails because it has none? https://github.com/nextcloud-snap/nextcloud-snap/actions/runs/6270627042/job/17028845113#step:4:12

kyrofa commented:
![Screenshot from 2023-09-26 11-07-56](https://github.com/nextcloud-snap/nextcloud-snap/assets/946674/8a83ac20-b436-4d1a-bd7b-a9da6352a511)

Man, finally :partying_face:

Pilzinsel64 commented:
Alright forks! I refreshed my private prod to this beta channel and configured notify_push to use the url via https now. Looks like notify_push is working as expected and without any warnings or errors anymore.

I showed the browser (files app open), test client (the window below) and the self-test of the notify_push client (the upper window).

Also I tested this (well, only the browser, not via the test client) outside of my network via the Tor Browser. Also here it seems to work fine.

![grafik](https://github.com/nextcloud-snap/nextcloud-snap/assets/23138465/5cfc2a76-6b6e-4937-b794-3ce16dc20c53)

pachulo commented:
I cannot really test this on my instance @Pilzinsel64, but it looks really nice!
What I feel is missing is some updates to the README, explaining how all this works together and what steps need to be done, WDYT?

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -6,6 +6,7 @@
 DEFAULT_HTTP_PORT="80"
 DEFAULT_HTTPS_PORT="443"
 DEFAULT_HTTP_COMPRESSION="false"
+DEFAULT_HTTP_NOTIFY_PUSH_ENABLED="false"
```
```suggestion
DEFAULT_HTTP_NOTIFY_PUSH="false"
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+
+apache_get_notify_push_enabled()
+{
+	compression="$(snapctl get http.notify-push)"
```
`compression`?

Pilzinsel64 replied:
Is that a local variable?

Pilzinsel64 replied:
I renamend it to `notify_push` in that method, looks like it's not used somewhere else.

Pilzinsel64 commented:
Arg, well, I did it now finally. But the clean up I still need to learn a bit. Problem here were that GitHub at "Code" tab always think I am on upstream/master and not develop. I only should rebase here in the PR, never on the repo itself. xD

pachulo commented on src/apache/bin/httpd-wrapper:
```diff
@@ -39,6 +39,13 @@ if debug_mode_enabled; then
 fi
 
+if apache_notify_push_enabled; then
+	echo "notify_push is enabled"
```
```suggestion
	echo "notify_push reverse proxy is enabled"
```

pachulo commented on src/apache/bin/httpd-wrapper:
```diff
@@ -39,6 +39,13 @@ if debug_mode_enabled; then
+	echo "notify_push is enabled"
+	params="$params -DEnableNotifyPush"
+else
+	echo "notify_push is disabled"
```
```suggestion
	echo "notify_push reverse proxy is disabled"
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -6,6 +6,7 @@
 DEFAULT_HTTP_PORT="80"
 DEFAULT_HTTPS_PORT="443"
 DEFAULT_HTTP_COMPRESSION="false"
+DEFAULT_HTTP_NOTIFY_PUSH="false"
```
```suggestion
DEFAULT_HTTP_NOTIFY_PUSH_REVERSE_PROXY="false"
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
 	snapctl set private.http.compression="$1"
 }
+
+apache_notify_push_enabled()
```
```suggestion
apache_notify_push_reverse_proxy_enabled()
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+
+apache_notify_push_enabled()
+{
+	[ "$(apache_get_notify_push_enabled)" = "true" ]
```
```suggestion
	[ "$(apache_get_notify_push_reverse_proxy_enabled)" = "true" ]
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+	[ "$(apache_get_notify_push_enabled)" = "true" ]
+}
+
+apache_enable_notify_push_enabled()
```
```suggestion
apache_enable_notify_push_reverse_proxy_enabled()
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+
+apache_enable_notify_push_enabled()
+{
+	_apache_set_notify_push_enabled "true"
```
```suggestion
	_apache_set_notify_push_reverse_proxy_enabled "true"
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+	_apache_set_notify_push_enabled "true"
+}
+
+apache_disable_notify_push_enabled()
```
```suggestion
apache_disable_notify_push_reverse_proxy_enabled()
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+
+apache_disable_notify_push_enabled()
+{
+	_apache_set_notify_push_enabled "false"
```
```suggestion
	_apache_set_notify_push_reverse_proxy_enabled "false"
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+	_apache_set_notify_push_enabled "false"
+}
+
+apache_notify_push_enabled_has_changed()
```
```suggestion
apache_notify_push_reverse_proxy_enabled_has_changed()
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+
+apache_notify_push_enabled_has_changed()
+{
+	[ "$(apache_get_notify_push_enabled)" != "$(_apache_get_previous_notify_push_enabled)" ]
```
```suggestion
	[ "$(apache_get_notify_push_reverse_proxy_enabled)" != "$(_apache_get_previous_notify_push_reverse_proxy_enabled)" ]
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+	[ "$(apache_get_notify_push_enabled)" != "$(_apache_get_previous_notify_push_enabled)" ]
+}
+
+apache_get_notify_push_enabled()
```
```suggestion
apache_get_notify_push_reverse_proxy_enabled()
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+
+apache_get_notify_push_enabled()
+{
+	notify_push="$(snapctl get http.notify-push)"
```
```suggestion
	notify_push_reverse_proxy="$(snapctl get http.notify-push-reverse-proxy)"
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+{
+	notify_push="$(snapctl get http.notify-push)"
+
+	if [ -z "$notify_push" ]; then
```
```suggestion
	if [ -z "$notify_push_reverse_proxy" ]; then
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+	notify_push="$(snapctl get http.notify-push)"
+
+	if [ -z "$notify_push" ]; then
+		notify_push="$DEFAULT_HTTP_NOTIFY_PUSH"
```
```suggestion
		notify_push_reverse_proxy="$DEFAULT_HTTP_NOTIFY_PUSH_REVERSE_PROXY"
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+
+	if [ -z "$notify_push" ]; then
+		notify_push="$DEFAULT_HTTP_NOTIFY_PUSH"
+		_apache_set_notify_push_enabled "$notify_push"
```
```suggestion
		_apache_set_notify_push_reverse_proxy_enabled "$notify_push_reverse_proxy"
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+		_apache_set_notify_push_enabled "$notify_push"
+	fi
+
+	echo "$notify_push"
```
```suggestion
	echo "$notify_push_reverse_proxy"
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+	echo "$notify_push"
+}
+
+_apache_set_notify_push_enabled()
```
```suggestion
_apache_set_notify_push_reverse_proxy_enabled()
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+
+_apache_set_notify_push_enabled()
+{
+	snapctl set http.notify-push="$1"
```
```suggestion
	snapctl set http.notify-push-reverse-proxy="$1"
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+_apache_set_notify_push_enabled()
+{
+	snapctl set http.notify-push="$1"
+	_apache_set_previous_notify_push_enabled "$1"
```
```suggestion
	_apache_set_previous_notify_push_reverse_proxy_enabled "$1"
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+	_apache_set_previous_notify_push_enabled "$1"
+}
+
+_apache_get_previous_notify_push_enabled()
```
```suggestion
_apache_get_previous_notify_push_reverse_proxy_enabled()
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+
+_apache_get_previous_notify_push_enabled()
+{
+	snapctl get private.http.notify-push
```
```suggestion
	snapctl get private.http.notify-push-reverse-proxy
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+	snapctl get private.http.notify-push
+}
+
+_apache_set_previous_notify_push_enabled()
```
```suggestion
_apache_set_previous_notify_push_reverse_proxy_enabled()
```

pachulo commented on src/apache/utilities/apache-utilities:
```diff
@@ -153,3 +154,51 @@ _apache_set_previous_http_compression()
+
+_apache_set_previous_notify_push_enabled()
+{
+	snapctl set private.http.notify-push="$1"
```
```suggestion
	snapctl set private.http.notify-push-reverse-proxy="$1"
```

pachulo commented on src/hooks/bin/configure:
```diff
@@ -19,6 +19,10 @@
 #   Whether or not HTTP compression should be enabled. Valid values are 'true'
 #   and 'false'.
 #
+# - http.notify-push (string)
```
```suggestion
# - http.notify-push-reverse-proxy (string)
```

pachulo commented on src/hooks/bin/configure:
```diff
@@ -156,8 +160,27 @@ handle_http_compression()
 	snapctl restart nextcloud.apache
 }
 
+handle_notify_push()
```
```suggestion
handle_notify_push_reverse_proxy()
```

pachulo commented on src/hooks/bin/configure:
```diff
@@ -156,8 +160,27 @@ handle_http_compression()
+handle_notify_push()
+{
+	# If no changes were requested, then there's nothing to do here.
+	if ! apache_notify_push_enabled_has_changed; then
```
```suggestion
	if ! apache_notify_push_reverse_proxy_enabled_has_changed; then
```

pachulo commented on src/hooks/bin/configure:
```diff
@@ -156,8 +160,27 @@ handle_http_compression()
+	fi
+
+	# Validate input
+	value="$(apache_get_notify_push_enabled)"
```
```suggestion
	value="$(apache_get_notify_push_reverse_proxy_enabled)"
```

pachulo commented on src/hooks/bin/configure:
```diff
@@ -156,8 +160,27 @@ handle_http_compression()
+	# Validate input
+	value="$(apache_get_notify_push_enabled)"
+	if [ "$value" != "true" ] && [ "$value" != "false" ]; then
+		echo "http.notify-push must be either 'true' or 'false'"
```
```suggestion
		echo "http.notify-push-reverse-proxy must be either 'true' or 'false'"
```

pachulo commented on src/hooks/bin/configure:
```diff
@@ -156,8 +160,27 @@ handle_http_compression()
 handle_cronjob_interval
 handle_mode
 handle_http_compression
+handle_notify_push
```
```suggestion
handle_notify_push_reverse_proxy
```

pachulo commented on src/apache/bin/httpd-wrapper:
```diff
@@ -39,6 +39,13 @@ if debug_mode_enabled; then
 
+if apache_notify_push_enabled; then
+	echo "notify_push is enabled"
+	params="$params -DEnableNotifyPush"
```
```suggestion
	params="$params -DEnableNotifyPushReverseProxy"
```

pachulo commented on src/apache/conf/httpd.conf:
```diff
@@ -206,6 +206,13 @@ ServerSignature Off
 </IfDefine>
 
+# Enable Reverse Proxy for notify_push
+<IfDefine EnableNotifyPush>
```
```suggestion
<IfDefine EnableNotifyPushReverseProxy>
```

pachulo commented on src/apache/conf/httpd.conf:
```diff
@@ -206,6 +206,13 @@ ServerSignature Off
     ServerSignature On
 </IfDefine>
 
+# Enable Reverse Proxy for notify_push
```
```suggestion
# Enable notify_push reverse proxy
```

pachulo commented on src/hooks/bin/configure:
```diff
@@ -19,6 +19,10 @@
 #   and 'false'.
 #
+# - http.notify-push (string)
+#   Whether or not the Reverse Proxy for notify_push should be enabled. Valid
```
```suggestion
#   Whether or not the notify_push reverse proxy should be enabled. Valid
```

Pilzinsel64 commented:
@pachulo Read to review again.

Should I cleanup again? If yes, I would do if everything is fine because it needs another 5 to 10 minutes for me. xD

Pilzinsel64 commented:
> I cannot really test this on my instance @Pilzinsel64, but it looks really nice! What I feel is missing is some updates to the README, explaining how all this works together and what steps need to be done, WDYT?

@pachulo Thank you! I'm sorry, I completely missed this comment...
Of course! But not sure if the Readme is the correct place here? There is already a Wiki entry that just needs to be adjusted when this has been merged. Not sure how I could explain in the README.md. 🤔 
See the wiki entry: https://github.com/nextcloud-snap/nextcloud-snap/wiki/FAQ's#q-how-to-install-files-hpb-client-push

pachulo commented on README.md:
```diff
@@ -126,6 +127,22 @@ To disable it, run:
+
+#### Reverse Proxy for Files High Performance Backeend
+
+This option simply enables the reverse proxy configuration mentoined in [Client Push Readme](https://github.com/nextcloud/notify_push#apache) that is required to setup the `notify_push` componente. Read more [at our wiki](https://github.com/nextcloud-snap/nextcloud-snap/wiki/FAQ's#q-how-to-install-files-hpb-client-push)!
```
```suggestion
This option simply enables the reverse proxy configuration mentioned in [Client Push Readme](https://github.com/nextcloud/notify_push#apache) that is required to setup the `notify_push` component. Read more [at our wiki](https://github.com/nextcloud-snap/nextcloud-snap/wiki/FAQ's#q-how-to-install-files-hpb-client-push)!
```

pachulo commented:
This now looks OK to me, WDYT @kyrofa ?

Pilzinsel64 commented:
FYI: My prod is still on `latest/beta/pr-2490` and notify_push via the reverse proxy still works fine, even after updates.

LMRW commented:
Watching that build and excited for this update! Thanks everyone. NC Snap is the best.

pachulo commented:
> Watching that build and excited for this update! Thanks everyone. NC Snap is the best.

Did you test it @LMRW?

LMRW commented:
> > Watching that build and excited for this update! Thanks everyone. NC Snap is the best.
> 
> Did you test it @LMRW?

I'd be happy to try - I assume I can test with `sudo snap install nextcloud --beta`?

Pilzinsel64 commented:
@LMRW Install with `sudo snap install nextckoud --channel=latest/beta/pr-2490` or refresh with `sudo snap refresh nextcloud --channel=latest/beta/pr-2490`.

pachulo commented:
> Install with `sudo snap install nextckoud --channel=latest/beta/pr-2490` or refresh with `sudo snap refresh nextcloud --channel=latest/beta/pr-2490`.

One more thing @LMRW : those will only work if your server is `amd64` based.

LMRW commented:
> > Install with `sudo snap install nextckoud --channel=latest/beta/pr-2490` or refresh with `sudo snap refresh nextcloud --channel=latest/beta/pr-2490`.
> 
> One more thing @LMRW : those will only work if your server is `amd64` based.

I just came back to mention this and saw your message - i'm afraid my device is arm64 based.

Please as soon as there is an arm64 build let me know and I will do all I can to help test. I am familiar with how notify_push should function so I will be able to test thoroughly.

Pilzinsel64 commented:
@kyrofa Would you please kick builds for all architectures?

kyrofa commented:
No, I just use `snapcraft remote-build`. It was broken and needed hand-holding while we were stuck on snapcraft 6, but it should work by default now.

LMRW commented:
Id like to learn more about your processes to help more but cant find any dev-centric documentation on how to get involved, try betas etc. Is that something I can help with? Possibly even add a few pages to the Wiki for you once I know?

pachulo commented:
OK, there are now builds of this for arm64 (and all the other archs), so you should be able to test it @LMRW !
By the way, nevermind the version on the name of the installed snap: check that you have the correct one (27.1.3 right now) on nextcloud itselft.
I will also try to test this, but I don't have that much time :sweat:

LMRW commented:
Apologies for the delay, I still fully intend to test this. As I only had one production device my plan is to setup a new VM for testing this - I aim to do this during coming week for you. Sorry for delay.

Pilzinsel64 commented:
Is it likely we'll get this ready for the next update in a week? A lot more people are able to test then via latest/beta.

Pilzinsel64 commented:
Err... my snap refreshed last night but there was no update?? The commits with the reverse proxy are not included anymore! What happened here?

![grafik](https://github.com/nextcloud-snap/nextcloud-snap/assets/23138465/4a966641-3c51-47fc-a10d-7cc64a736671)

```sh
ubuntu@dcloud:~$ sudo snap info nextcloud
...
snap-id:      njObIbGQEaVx1H4nyWxchk1i8opy4h54
tracking:     latest/beta/pr-2490
refresh-date: yesterday at 22:01 UTC
...
installed:          27.1.3snap1+git2.7040ce3            (38881) 313MB -
```

kyrofa commented:
Snapcraft branches are temporary: they expire after 30 days and close themselves, so you fell back to the beta channel.

Pilzinsel64 commented:
> Snapcraft branches are temporary: they expire after 30 days and close themselves, so you fell back to the beta channel. You can resurrect it by releasing again, e.g. updating this git branch.

Ah, so 30 days after migration, right? Because it has been updated the last week.

kyrofa commented:
> Ah, so 30 days after creation, right? Because it has been updated the last week.

No it _should_ be 30 days after the last _update_... but who knows what those people are doing these days.

> And I wonder why snap info still says the temporary branche instead of latest/beta again.

You can still track a closed channel, but closed channels inherit the next-most-stable channel.

Pilzinsel64 commented:
Within the next days I will downgrade to http connection with extra port again and switch back to `latest/beta` according to be able to test nc28 via `28/edge`. That means I will not be able to test this PR anymore, sorry.

From my side, this PR works 100%ly as expected and I didn't see any downside over the last two/three months.

pachulo commented on README.md:
```diff
@@ -126,6 +127,22 @@ To disable it, run:
+
+    $ sudo snap set nextcloud http.notify-push-reverse-proxy=false
+
+Note: You still need to setup `notify_push` yourself. This option only enables the reverse proxy, as the apache configuration is read-only.
```
No, it's more complicated than that: https://github.com/nextcloud-snap/nextcloud-snap/wiki/Configure-HPB-client-push-for-Nextcloud-snap
This config that @Pilzinsel64 created "only" allows for it to work though HTTPS, as we couldn't really think of a way to automate the other steps

kyrofa commented on src/apache/conf/httpd.conf:
```diff
@@ -206,6 +206,13 @@ ServerSignature Off
 
+# Enable notify_push reverse proxy
+<IfDefine EnableNotifyPushReverseProxy>
+    ProxyPass /push/ws ws://127.0.0.1:7867/ws
+    ProxyPass /push/ http://127.0.0.1:7867/
+    ProxyPassReverse /push/ http://127.0.0.1:7867/
```
Quickly perusing the notify_push README leads me to believe this port is entirely up to the user, right? Looks like maybe they need to setup a service that listens on this port? Is it worth making that configurable? Also, if we're going to restrict this to localhost, we might also consider restricting it to a unix socket instead of a port.

pachulo replied:
> Quickly perusing the notify_push README leads me to believe this port is entirely up to the user, right?

I think so

> Looks like maybe they need to setup a service that listens on this port?

Totally. See https://github.com/nextcloud-snap/nextcloud-snap/wiki/Configure-HPB-client-push-for-Nextcloud-snap#b-install-system-deamon

> Is it worth making that configurable?

I don't really see the benefit of making it configurable, to be honest

> If not, if we're going to restrict this to localhost, we might also consider restricting it to a unix socket instead of a port.

That's a great idea, but I don't know how it can be done...

Pilzinsel64 replied:
Yeah, didn't know socket path is possible... But how to get sich a socket path and how to setup that in Apache for the lines above? Never did that and never seen that.

Pilzinsel64 replied:
Something like that? Do I need to do something else with it or just use it?
`redis+unix:///tmp/snap-private-tmp/snap.nextcloud/tmp/sockets/notify_push.sock`

pachulo replied:
I'm sorry @Pilzinsel64 , I know that we are also using a socket for the nextcloud <-> MySQL communication, but I don't really know the details of how it works, or how we should use one in this case...

kyrofa replied:
Yeah I would try copying what we do for the PHP reverse proxy. Operating off of memory, I believe it looks like what you have there, @Pilzinsel64. Give it a shot and see if it works!

Pilzinsel64 replied:
Oh dear... two hours of try-and-error later I got a configuration that worked on another production instance (without snap). Let's see if the configuration works here too, because I had trouble getting apache to see my socket in /temp/... while my tests.

The only thing I really wonder is how the paths now should look like. For Apache **inside** the snap I now use the "fake" path to the socket and for the [service configuration](https://github.com/nextcloud-snap/nextcloud-snap/wiki/Configure-HPB-client-push-for-Nextcloud-snap/_compare/19ff9c0eb2e32d1515e7ada0b8c36ad1ee0f6b68...5442930a348d36211d83d34156fc47fc762073da) **outside** the snap I used the "real" path.

@kyrofa can you please double check the socket paths? Should the path itself correct like that?

As soon as the build successed I will try the snap on my test VM.

kyrofa commented on src/apache/conf/httpd.conf:
```diff
@@ -206,6 +206,13 @@ ServerSignature Off
 
+# Enable notify_push reverse proxy
+<IfDefine EnableNotifyPushReverseProxy>
+    ProxyPass /push/ws unix:/tmp/sockets/notify_push.sock|ws://localhost/ws
+    ProxyPass /push/ unix:/tmp/sockets/notify_push.sock|http://localhost/
+    ProxyPassReverse /push/ unix:/tmp/sockets/notify_push.sock|http://localhost/
```
Yes these paths are what I would expect within the snap :+1: .

pachulo commented:
> Hi! I'm currently making new builds for this, will upload them when they are available and I will try to test it in my production instance this next week.

Damm, that took long, but I was finally able to upload the latest builds to the `latest/beta/pr-2490` channel

pachulo commented:
OK, I hit the first problem: when trying to upgrade my instance, which is on `26.0.10snap1`, I get this error:
```
$ sudo snap refresh nextcloud --channel latest/beta/pr-2490
2024-01-15T22:01:48Z INFO Waiting for "snap.nextcloud.mysql.service" to stop.
2024-01-15T22:01:55Z INFO Waiting for "snap.nextcloud.apache.service" to stop.
2024-01-15T22:01:57Z INFO Waiting for "snap.nextcloud.redis-server.service" to stop.
error: cannot perform the following tasks:
- Run post-refresh hook of "nextcloud" snap if present (run hook "post-refresh": Nextcloud doesn't support downgrades, but you can revert. See 'snap revert -h' for details.)
```

I've checked with a fresh install and indeed the `latest/beta/pr-2490` channel installs a nextcloud version 27.1.4.
What's going on @kyrofa ?
Might it be that the snaps I built from @Pilzinsel64 [fork](https://github.com/Pilzinsel64/nextcloud-snap/tags) are using an old tag?
`nextcloud_25.0.2snap1+git168.2aca295_ARCH.snap`

Pilzinsel64 commented:
@pachulo I rebased once again from upstream via `git rebase --interactive upstream/develop` after adding this repo as remote: `git remote add upstream https://github.com/nextcloud-snap/nextcloud-snap.git`

This also fetched the tags and I hope they are also present here now when the builds has finished and uploaded.

pachulo commented:
Hi @Pilzinsel64 ! I don't think you imported the tags correctly: latest in your fork is still [25.0.2snap1](https://github.com/Pilzinsel64/nextcloud-snap/releases/tag/25.0.2snap1).

I think that you should be able to do so using `git fetch upstream --tags` and then `git push --tags`

> (Sorry for the long delay, had not in mind I could do that.)

My delay was longer, I win :wink:

Pilzinsel64 commented:
Ah, it already fetched the tags corrects, but I had to push them with --tags. Forgot that completely that it is required to explicitely tell git that. Thanks!

However, I tried to rebase once again but without success. It won't force-push this time. 🤔 You may need to kick off the builds yourself. @pachulo 
![grafik](https://github.com/nextcloud-snap/nextcloud-snap/assets/23138465/9ed9b2a5-7df4-444d-8f89-6cf1406bd6fe)

pachulo commented:
OK, I was finally able to test this, but I found some problems:
1. I had to add the [`trusted_proxies`](https://docs.nextcloud.com/server/27/admin_manual/configuration_server/config_sample_php_parameters.html?highlight=overwrite%20cli%20url#:~:text=List%20of%20trusted%20proxy%20servers) key with value `127.0.0.1` in my `/var/snap/nextcloud/current/nextcloud/config/config.php` file.
2. I had to update the [`overwrite.cli.url`](https://docs.nextcloud.com/server/27/admin_manual/configuration_server/config_sample_php_parameters.html?highlight=overwrite%20cli%20url#:~:text=%27overwrite.cli.url%27%20%3D%3E%20%27%27%2C) key with my main Nextcloud instance domain, as it was configured with its internal IP.

I will do some more tests latter this week!

Pilzinsel64 commented:
> * I had to add the [`trusted_proxies`](https://docs.nextcloud.com/server/27/admin_manual/configuration_server/config_sample_php_parameters.html?highlight=overwrite%20cli%20url#:~:text=List%20of%20trusted%20proxy%20servers) key with value `127.0.0.1` in my `/var/snap/nextcloud/current/nextcloud/config/config.php` file.
> 
> * I also had to update the [`overwrite.cli.url`](https://docs.nextcloud.com/server/27/admin_manual/configuration_server/config_sample_php_parameters.html?highlight=overwrite%20cli%20url#:~:text=%27overwrite.cli.url%27%20%3D%3E%20%27%27%2C) key with my main Nextcloud instance domain, as it was configured with its internal IP (but I don't really know why).

Yeah true, that's common. Well, I usually had to add my public IP as trusted proxy (or domain) and not to 127.0.0.1. But that might depend on your setup. But can't really remember, absolutely forgot about. 

Will take a look on my config on all my instances and add that to the wiki.

pachulo commented:
I've updated the [wiki page](https://github.com/nextcloud-snap/nextcloud-snap/wiki/Configure-HPB-client-push-for-Nextcloud-snap) with uninstall instructions.

What else is missing here for this being mergeable @kyrofa ?

I guess we still need to answer some of @Pilzinsel64 questions from [the original message](https://github.com/nextcloud-snap/nextcloud-snap/pull/2490#issue-1839567145)...

pachulo commented:
After a restart the systemd unit notify_push was in failed state:
```
$ sudo systemctl status notify_push
● notify_push.service - Push daemon for Nextcloud clients 
     Loaded: loaded (/etc/systemd/system/notify_push.service; enabled; vendor preset: enabled)
     Active: failed (Result: exit-code) since Sat 2024-01-27 14:40:41 UTC; 3min 54s ago
    Process: 801 ExecStart=/var/snap/nextcloud/current/nextcloud/extra-apps/notify_push/bin/aarch64/notify_push />
   Main PID: 801 (code=exited, status=1/FAILURE)
                            
Jan 27 14:40:40 raspi3 systemd[1]: Started Push daemon for Nextcloud clients.
Jan 27 14:40:41 raspi3 notify_push[801]: Error:   × Failed to connect to database
Jan 27 14:40:41 raspi3 notify_push[801]:   ├─▶ error communicating with database: No such file or directory (>
Jan 27 14:40:41 raspi3 notify_push[801]:   │   2)
Jan 27 14:40:41 raspi3 notify_push[801]:   ╰─▶ No such file or directory (os error 2)
Jan 27 14:40:41 raspi3 systemd[1]: notify_push.service: Main process exited, code=exited, status=1/FAILURE
Jan 27 14:40:41 raspi3 systemd[1]: notify_push.service: Failed with result 'exit-code'.
```


so I guess that we will need to modify the template, to make it depend on some of the nextcloud snap units...

Pilzinsel64 commented:
> After a restart the systemd unit notify_push was in failed state:

@pachulo I never had that problem. And I still don't have it. I can restart that service without problem. Could it be that the channel has been removed again? I'm going to rebase once again.

kyrofa commented:
> What else is missing here for this being mergeable @kyrofa ?

As far as I can tell, we can't really add tests for this because using it requires more infrastructure. I won't lie and say I love this, but I can see the usefulness behind it, and this is literally the only way to pull it off. As long as you guys are here to maintain it and it's off by default, feel free to merge once you all are happy with it.

Pilzinsel64 commented:
> That's exactly what I would like to avoid by default using some systemd magic. I will look for solutions whenever I can

Well, if you want a bit more time we can also do Followup to optimize it later. This is the way how it works oficially by Nextcloud (without the tweaks but to be honest, I need that on all instances). It's not our fault and should fixed by Nextcloud by just trying to reconnect once every minute until it success. I tent to make an issue at the notify_push repo if you agree.

Pilzinsel64 commented:
@pachulo I've added another line to the systemd file, see the wiki page: https://github.com/nextcloud-snap/nextcloud-snap/wiki/Configure-HPB-client-push-for-Nextcloud-snap#b-create-the-systemd-unit

I used the base from this issue comment: https://github.com/nextcloud/notify_push/issues/194#issuecomment-1331513813

This should solve our problem. I already added this to my prod instance.

pachulo commented:
> @pachulo I've added another line to the systemd file to defines after wich services the notify_push service should start, see the wiki page: https://github.com/nextcloud-snap/nextcloud-snap/wiki/Configure-HPB-client-push-for-Nextcloud-snap#b-create-the-systemd-unit
> 
> I used the base from this issue comment: https://github.com/nextcloud/notify_push/issues/194#issuecomment-1331513813
> 
> This should solve our problem. I already added this to my prod instance.

Yes, that's what I was talking about. The only thing is that it _does not actually solve it_, because the units are started in parallel and I suspect that the `notify_push` one needs for everything to be up and running before being able to start.

Anyway, I've found that adding this to the `[Service]` stanza makes it start OK after a reboot:
```
ExecStartPre=/bin/bash -c 'while /bin/systemctl is-active --quiet snap.nextcloud.nextcloud-fixer.service; do sleep 10; done'
TimeoutStartSec=15m
```
It just checks that the `snap.nextcloud.nextcloud-fixer.service` unit has finished doing its thing, to then try to start the `notify_push` unit. It timeouts after 15 minutes, which is quite arbitrary....

Opinions @Pilzinsel64 & @kyrofa ?
Not the most elegant thing in the world, but to me this is a problem with the functionality itself, not the snap, and we would be offering something that is at least as robust, if not more, than the standard solution.

Pilzinsel64 commented:
@pachulo I agree. I added this to the wiki and changed the file on my prod server. Let's see how well it works over the time. :)

I also rebased and merged the commits.

pachulo commented:
I've updated the wiki with the exact options that I've used. Let's see if people don't hit too many problems with our config, or if the notify_push crew make any official change to their template.

pachulo commented:
Thanks A LOT for your hard work and patience @Pilzinsel64 ! Your passion is really appreciated!
